### PR TITLE
Add format rebase bot command

### PR DESCRIPTION
**Pull request by upsj**
Instead of filling the git history with ` format `  commits, this PR adds a ` format-rebase!`  command that formats each individual commit instead by rewriting the history.

### CODE REVIEW

**Comment by pratikvn** [comment]
I am not too sure of this. How would this work ? Would it rewrite each of the commits ? That might make things unnecessarily complicated with different commits on your local system and on github. I feel that this would be more error-prone.

**Comment by upsj** [comment]
Yes, like `rebase!` this rewrites the history for each individual commit, so you need to `git fetch && git reset --hard origin/<branch>` to get the new state locally. An alternative would be to run the command from the shell script manually.

To elaborate a bit on my thoughts for this: It is helpful for all kinds of operations (`git blame`, `git bisect`, ...) if most if not all intermediate states of `develop` are in a clean state. This is especially relevant for the few long-lived branches we have (batch, MPI, block-ILU, ...), since they can be hundreds of commits ahead of the code.

**Review by tcojean — APPROVED** [review]
In general LGTM, I like the idea.

Do you have an example of it working in your playground?

**Comment by upsj** [comment]
@tcojean I made a few updates in the playground (especially storing the individual diffs as artifacts of the job, to be able to roll back any potential breakages), but in general you can check it out for yourself, yes. You should have access: https://github.com/upsj/ginkgo-bot-playground

File: .github/format-rebase.sh [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+pushd dev_tools/scripts
+cp add_license.sh format_header.sh update_ginkgo_header.sh /tmp
+popd
+
+bot_delete_comments_matching "Error: Rebase failed"
+
+DIFF_COMMAND="git diff --name-only --no-renames --diff-filter=AM HEAD~ | grep -E '$EXTENSION_REGEX'"
+
+# do the formatting rebase
+git rebase --empty=drop --no-keep-empty \
```
[the remark this answers is not in the document]
Comment (upsj) [reply .github/format-rebase.sh:29]: I don't think that's a good idea - rebasing the complete topology of a branch involving merges is pretty complex and loses all connection to the original branches. `batch-develop` is enough of a special case that we can do that by hand.

File: .github/workflows/bot-pr-comment.yml [hunk 2]
```diff
@@ -63,3 +63,18 @@ jobs:
...
+    - name: Checkout the latest code
+      uses: actions/checkout@v2
+      with:
+        ref: develop
+        fetch-depth: 0
+        persist-credentials: false
+    - name: Automatic Formatting Rebase
+      env:
+        GITHUB_TOKEN: ${{ secrets.BOT_TOKEN }}
+      run: cp .github/format-rebase.sh /tmp && /tmp/format-rebase.sh
```
[the remark this answers is not in the document]
Comment (upsj) [reply .github/workflows/bot-pr-comment.yml:80]: the file may be overwritten when checking out a different state from the repo. Copying it avoids the resulting race conditions between bash reading the file and git updating it.

**Pull request merged**